### ISSUE TITLE
Move the Kubernetes nodes from EC2 to EKS Fargate

### ISSUE THREAD

**Issue by jimleroyer** [cds-snc/notification-planning-core#55]
**Current estimate of 3 is for writing the ADR only.**

# Description

An a SRE specialist in GCNotify,
I want to move the Kubernetes nodes from EC2 provisioning to EKS Fargate,
So that I can run on serverless architecture and scale easier [1]

[1] [AWS Elastic Kubernetes Service (EKS): EC2 vs Fargate](https://www.mobilise.cloud/aws-elastic-kubernetes-service-eks-ec2-vs-fargate/)

> What are the main benefits of Fargate?
>
> * Not having to worry about scaling EC2 nodes – you provision more pods on Fargate, and you have unlimited compute
> * Right-Sizing – AWS Fargate matches your workloads requirements with a set of pre-configured CPU & Memory based price points (more on this later)
> * Secure isolation by design – individual EKS pods each run in their own dedicated kernel runtime environment and do not share resources with any other pod

## WHY are we building?

* Scale automatically without having to reconfigure the nodes.
* Reduce required maintenance on the long run.

## WHAT are we building?

Moving from EC2 provisioning to EKS Fargate.

## VALUE created by our solution

* Scalability. 
* Less time spent on maintenance and scalability in the future.

# Acceptance Criteria** (Definition of done) 

- [ ] Explore the architecture change within an ADR first. Answer questions such as compatibility issues we might encounter (e.g. Daemonsets support, AWS Container Insights) and if we expect a decrease or increase in cost.

# QA Steps

- [ ] Review ADR with Notify developers and SRE team.

# Additional Info

## Drawbacks

> * Without EKS Fargate Spot, EC2 is cheaper.
> * Daemonsets are not supported on Fargate. This means you will need to re-engineer solutions (such as pre-made HELM charts) to use sidecars in your containers instead.
> * Privileged containers are not supported (although this is generally best practice)
> * GPUs are currently not available in Fargate.
> * Only Private Subnets are supported (generally best practice)
> * AWS Container Insights is currently unsupported on EKS Fargate

https://www.mobilise.cloud/aws-elastic-kubernetes-service-eks-ec2-vs-fargate

## Rationale

This task is for scaling our kubernetes nodes (hosting our celery workers, i.e. power to process and send notifications) more automatically. We'd have to adjust our EC2 machine configurations at the moment but if we move to Fargate, with the serverless model, we could scale with less effort and less need to maintain our environment as well (manually maintained EC2 environment versus serverless with an environment provided for you).

**Comment by ben851** [comment]
Actually investigating this, looks like auto eks is better for our use case. I broke dev trying to implement it yesterday on the existing cluster. Recreating dev now, and then will look into adding a new auto eks cluster that we will migrate to from the old one.

**Comment by jimleroyer** [comment]
Duplicate of #705. Keeping this one and killing the other.

**Comment by schancase** [comment]
Note - Sarah will parent this under our new O1KR2

**Comment by schancase** [comment]
This is now parented under Epic #828, which is under our O1KR2

**Comment by ben851** [comment]
EKS Migration Spike: Fargate vs. Auto Mode vs. Status Quo
EKS Fargate — Not Viable
EKS Fargate was evaluated but has fundamental incompatibilities with our current architecture.

DaemonSets are not supported. A significant portion of our cluster's observability and security stack relies on DaemonSets running on every node. Migrating to Fargate would require replacing all of the following with per-pod sidecars or alternative architectures:

CloudWatch Agent (aws-cloudwatch) — node-level metrics and log collection
X-Ray Daemon (xray-daemon) — distributed tracing
SigNoz Telemetry Collector (signoz-telemetry) — OTEL pipeline for traces and metrics
Fluent/FluentBit — log forwarding
Falco — runtime security monitoring (also requires privileged kernel access, which Fargate explicitly prohibits)
Prometheus Node Exporter — node-level metrics scraping
Adding sidecars to replace these would increase per-pod resource overhead across all workloads, introduce operational complexity, and drive up cost — particularly in a horizontally-scaled environment like ours.

Additional blockers:

No support for EBS volumes. Our storage classes are EBS-backed, and EFS would be a non-trivial migration for stateful workloads.
No equivalent of spot instances. Our Karpenter NodePools are configured to run on spot capacity (capacityTypes: ["spot"]) across four AZs (ca-central-1a/b/c/d), providing significant cost savings. Fargate has no comparable mechanism.
No privileged containers. Falco relies on kernel-level access which Fargate prohibits entirely, meaning our runtime security posture would degrade with no straightforward replacement.
Pod resource sizing is constrained to predefined vCPU/memory tiers, removing the fine-grained control we currently have via Karpenter.
Verdict: Fargate is not compatible with our current stack and would require a large-scale architectural overhaul for no clear benefit.

EKS Auto Mode — Marginal Benefit
EKS Auto Mode offloads node lifecycle management to AWS, effectively providing a managed version of what Karpenter does today. On the surface this is appealing, but in practice it offers limited value given our current setup.

We already have a well-tuned Karpenter configuration. Our self-managed Karpenter deployment includes:

Multiple NodePools with specific instance families (m5, r5) and sizes (2–4 vCPU)
A dedicated node pool for SigNoz to isolate its resource-intensive workloads
Spot-first capacity with configurable disruption budgets (nodes: 10%) and expiry policies (expireAfter: 720h)
Zone spread across all four ca-central-1 AZs for resilience
EKS Auto Mode would abstract away this fine-grained control. AWS manages the underlying EC2s more opaquely, which was a pain point we experienced previously with SSC — reduced visibility into the underlying infrastructure makes incident response and capacity debugging harder.

Additional considerations:

EKS Auto Mode has its own node class constraints that may not map directly to our existing NodePool configurations.
DaemonSets, node affinity rules, and workload placement logic would need to be re-validated against Auto Mode's behaviour.
There is a non-trivial migration effort with no compelling operational or cost benefit.
Verdict: EKS Auto Mode is essentially a managed wrapper around what we already do ourselves. The migration cost outweighs the benefit of having AWS manage Karpenter on our behalf.

Status Quo (EKS + Self-Managed Karpenter) — Recommended
Our current setup is well-understood, stable, and purpose-built for our workloads. Key strengths:

Full control over node provisioning, instance selection, and disruption policies
DaemonSet support enables our existing observability stack (CloudWatch, X-Ray, SigNoz, FluentBit, Falco) without architectural changes
Spot instance usage across multiple AZs provides meaningful cost savings
Goldilocks + VPA provide continuous right-sizing recommendations
Velero handles backup and restore for stateful workloads
Full node access for debugging and incident response
Conclusion: The spike recommends maintaining the status quo. Both alternatives introduce significant migration risk and complexity without resolving any current operational pain points. If node management overhead becomes a concern in the future, EKS Auto Mode could be revisited — but only after AWS matures its support for custom NodePool configurations.

**Comment by jimleroyer** [comment]
We got feedback from security team about this that they were planning for us to move to a serverless infrastructure:

> ing taken a look at that I think you are missing important context the SPIN on improving GC Cyber Security Health that was published August 2024 has a requirement that we install CCCS Host Based Sensors on all servers in cloud environments, ephemeral assets such as containers are exempt from this requirement.
> 
> We've kind of been sneaky and tried to get around this through the implementation of SCP's that prevent us from logging into the EC2 instances that your EKS is running on.
>
> However that is probably not a control that will stand up to scrutiny by a central governance body or anyone outside the org who pushes too hard on us regarding this requirement.
>
> We will try to justify but that mitigation was only ever in place with the expectation that Notify would migrate to either Fargate or Auto-EKS mode.
>
> Technically you should be building an AMI Pipeline for building CDS approved machine images that are sending telemetry to CCCS.
>
> This needs to be understood as a risk that Notify is taking on and you may find yourselves in a situation where you either need to implement the configurations identified in that SPIN on short notice, or be required to migrate to Auto-EKS/Fargate nodes.

Considering this, and thinking we don't want to go the fargate/eks route, the IaaS Route (AMI Pipeline) might be something worth it, as this could benefit other scenarios too. We created #874 to address that.

**Comment by schancase** [comment]
this task was re-parented into the Q3 version of the same epic
